frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java

```diff
@@ -11,65 +11,75 @@
-      }
+    Map<String, PlaceVisitModel> tripsMap = storage.get(placeVisit.placeId());
+
+    if (tripsMap == null) {
+      Map<String, PlaceVisitModel> newPlaceMap = new ConcurrentHashMap<>();
+      newPlaceMap.put(placeVisit.tripId(), placeVisit);
+      storage.put(placeVisit.placeId(), newPlaceMap);
+
+    } else if (tripsMap.get(placeVisit.tripId()) == null) {
+      tripsMap.put(placeVisit.tripId(), placeVisit);
+
+    } else {
+      throw new PlaceVisitAlreadyExistsException("PlaceVisit " + 
+        placeVisit.name() + " already exists for trip " + placeVisit.tripId());
     }
   }
 
   @Override
   public void removePlaceVisit(String tripId, String placeId) throws PlaceVisitNotFoundException {
-    if (storage.remove(placeId, tripId) == null) {
+    Map<String, PlaceVisitModel> tripsMap = storage.get(placeId);
+    if (tripsMap == null || tripsMap.remove(tripId) == null) {
       throw new PlaceVisitNotFoundException("PlaceVisit with id" + placeId + 
-      " not found for trip " + tripId);
+          " not found for trip " + tripId);
     }
   }
 
   @Override
   public PlaceVisitModel getPlaceVisit(String tripId, String placeId) throws PlaceVisitNotFoundException {
-    PlaceVisitModel place = storage.get(placeId, tripId);
-    if (place != null) {
-      return place;
+    Map<String, PlaceVisitModel> tripsMap = storage.get(placeId);
+    if (tripsMap == null || tripsMap.get(tripId) == null) {
+      throw new PlaceVisitNotFoundException("PlaceVisit with id" + placeId + 
+          " not found for trip " + tripId);
+    } else {
+      return tripsMap.get(tripId);
     }
-    throw new PlaceVisitNotFoundException("PlaceVisit with id" + placeId + 
-      " not found for trip " + tripId);
+    
   }
 
   @Override
   public boolean changePlaceVisitStatus(String tripId, String placeId, String newStatus) {
-    synchronized(storage) {
-      PlaceVisitModel place = storage.get(placeId, tripId);
-      if (place != null) {
-        PlaceVisitModel updatedPlace = place.toBuilder().setUserMark(newStatus).build();
-        storage.put(placeId, tripId, updatedPlace);
-        return true;
-      }
+    Map<String, PlaceVisitModel> tripsMap = storage.get(placeId);
+    if (tripsMap == null || tripsMap.get(tripId) == null) {
       return false;
+    } else {
+      PlaceVisitModel place = tripsMap.get(tripId);
+      PlaceVisitModel updatedPlace = place.toBuilder().setUserMark(newStatus).build();
+      tripsMap.put(tripId, updatedPlace);
+      return true;
     }
   }
 
   @Override
   public List<PlaceVisitModel> getTripPlaceVisits(String tripId) {
-    Map<String, PlaceVisitModel> placeIdPlaceMap = storage.column(tripId);
-    Collection<PlaceVisitModel> placeVisits = placeIdPlaceMap.values();
     List<PlaceVisitModel> tripPlaceVisits = new ArrayList<>();
-    for (PlaceVisitModel place: placeVisits) {
-      if (place.userMark().equals("must-see") || place.userMark().equals("if-time")) {
+    storage.forEach((placeId, tripIdMap) -> {
```

Comment:
Note that this suffers from the same (possible) issue as the trip storage model, unless you specifically want to use optimistic concurrency here (please see my comment in #31).

Either way, this code block is a good candidate for a stream-filter-collect flow, which might make the intent more obvious.